datafusion/src/optimizer/single_distinct_to_groupby.rs

```diff
@@ -0,0 +1,278 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! single distinct to group by optimizer rule
+
+use crate::error::Result;
+use crate::execution::context::ExecutionProps;
+use crate::logical_plan::{columnize_expr, DFSchema, Expr, LogicalPlan};
+use crate::optimizer::optimizer::OptimizerRule;
+use crate::optimizer::utils;
+use std::sync::Arc;
+
+/// single distinct to group by optimizer rule
+///   - Aggregation
+///          GROUP BY (k)
+///          F1(DISTINCT s)
+///       - X
+///
+///   into
+///
+///   - Aggregation
+///            GROUP BY (k)
+///            F1(s)
+///       - Aggregation
+///               GROUP BY (k, s)
+///            - X
+///   </pre>
+///   <p>
+pub struct SingleDistinctToGroupBy {}
+
+impl SingleDistinctToGroupBy {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
+    match plan {
+        LogicalPlan::Aggregate {
+            input,
+            aggr_expr,
+            schema,
+            group_expr,
+        } => {
+            if is_single_distinct_agg(plan) {
+                let mut all_group_args = Vec::with_capacity(group_expr.len());
+                all_group_args.append(&mut group_expr.clone());
+                // remove distinct and collection args
+                let new_aggr_expr = aggr_expr
+                    .iter()
+                    .map(|agg_expr| match agg_expr {
+                        Expr::AggregateFunction { fun, args, .. } => {
+                            all_group_args.append(&mut args.clone());
```

Comment:
Likewise here you can do something like this (and avoid a `mut`): 

```suggestion
                            all_group_args.extend(args.iter().cloned());
```